blog/models.py

```diff
@@ -127,6 +128,54 @@ class BlogArticlePage(MixinSeoFields, Page, MixinPageMethods, GoogleAdsMixin):
         StreamFieldPanel("body"),
     ]
 
+    @cached_property
+    def intro(self) -> str:
+        paragraph_text = self._get_paragraph_text_for_intro(MAX_BLOG_ARTICLE_INTRO_LENGTH)
+        intro_text = ""
+        if len(paragraph_text) > 0:
+            words = paragraph_text.split()
+            word_index = 0
+            new_text = words[word_index]
+            while len(new_text) < MAX_BLOG_ARTICLE_INTRO_LENGTH:
+                intro_text = new_text
+                word_index += 1
+                new_text = f"{intro_text} {words[word_index]}"
+        return intro_text
+
+    def _get_paragraph_text_for_intro(self, character_limit: int) -> str:
+        body_stream_child_index = self._get_index_of_next_nearest_paragraph()
+        paragraph_text = (
+            ""
+            if body_stream_child_index is None
+            else self._strip_rich_text_of_html_tags(
+                self.body[body_stream_child_index].value.source  # pylint:disable=unsubscriptable-object
+            )
+        )
+        while body_stream_child_index is not None and len(paragraph_text) < character_limit:
+            body_stream_child_index = self._get_index_of_next_nearest_paragraph(body_stream_child_index + 1)
+            if body_stream_child_index is not None:
+                next_paragraph_child_text = self.body[  # pylint:disable=unsubscriptable-object
+                    body_stream_child_index
+                ].value.source
+                paragraph_text += f" {self._strip_rich_text_of_html_tags(next_paragraph_child_text)}"
+        return paragraph_text
+
+    def _get_index_of_next_nearest_paragraph(self, body_stream_child_index: int = 0) -> Optional[int]:
+        while (
+            body_stream_child_index < len(self.body)
+            and self.body[body_stream_child_index].block.name != "paragraph"  # pylint:disable=unsubscriptable-object
+        ):
+            body_stream_child_index += 1
+        if body_stream_child_index >= len(self.body):
+            return None
+        return body_stream_child_index
```

Comment:
I did some refactor of this code because it was unreadable fore me and it was too long to write and probably when I would write everything even me won't be able to understand what I wanted to say so I decided to give you code instead of writing essays.
```suggestion
    @cached_property
    def intro(self) -> str:
        paragraph_text = self._get_paragraphs_text_for_intro(MAX_BLOG_ARTICLE_INTRO_LENGTH)
        if len(paragraph_text) == 0:
            return "Article intro not available."

        words_cycle = cycle(paragraph_text.split())
        return self._concatanate_intro_text_from_paragraphs_text(words_cycle, MAX_BLOG_ARTICLE_INTRO_LENGTH)

    def _concatanate_intro_text_from_paragraphs_text(self, words_cycle, character_limit):
        intro_text = ""
        new_text = next(words_cycle)

        while len(new_text) < character_limit:
            intro_text = new_text
            new_text = self._concatanate_strings(intro_text, words_cycle)
        return intro_text

    def _get_paragraphs_text_for_intro(self, character_limit: int) -> str:
        paragraphs: list = self._get_list_of_paragraphs()
        paragraphs_text = ""
        if len(paragraphs) == 0:
            return paragraphs_text

        paragraphs_cycle = cycle(paragraphs)

        while len(paragraphs_text) < character_limit:
            paragraphs_text = self._extract_paragraph_text_from_paragraph_block(paragraphs_cycle, paragraphs_text)

        return paragraphs_text

    def _get_list_of_paragraphs(self) -> list:
        return list(filter(lambda body_element: body_element.block.name == "paragraph", self.body))

    def _extract_paragraph_text_from_paragraph_block(self, paragraphs: cycle, text: str) -> str:
        space_between_texts = " "
        if len(text) == 0:
            text = self._strip_rich_text_of_html_tags(next(paragraphs).value.source)
        else:
            text += f"{space_between_texts}{self._strip_rich_text_of_html_tags(next(paragraphs).value.source)}"
        return text

    def _concatanate_strings(self, text: str, words: cycle) -> str:
        space_between_texts = " "
        text += f"{space_between_texts}{next(words)}"
        return text
```

This what I write here is passing all tests which are added, but for sure it also can be improved. Please tell me what you think about what I proposed here. I added more methods because for first look I lost a lot of time to understand what you wanted accomplished by code you wrote. If you want to discuss about it please contact me

Reply:
All right, I'll gladly take you up on that offer soon 👍